pkg/frontend/querymiddleware/astmapper/instant_splitting.go

```diff
@@ -393,37 +399,42 @@ func (i *instantSplitter) splitAndSquashCall(expr *parser.Call, stats *MapperSta
 
 // assertSplittableRangeInterval returns the range interval specified in the input expr and whether it is greater than
 // the configured split interval.
-func (i *instantSplitter) assertSplittableRangeInterval(expr parser.Expr) (rangeInterval time.Duration, canSplit bool) {
-	rangeInterval = getRangeInterval(expr)
+func (i *instantSplitter) assertSplittableRangeInterval(expr parser.Expr) (rangeInterval time.Duration, canSplit bool, err error) {
+	rangeIntervals := getRangeIntervals(expr)
+	if len(rangeIntervals) == 0 {
+		return time.Duration(0), false, nil
+	}
+	if len(rangeIntervals) > 1 {
+		return time.Duration(0), false, fmt.Errorf("found %d range intervals while expecting at most 1", len(rangeIntervals))
+	}
+
+	rangeInterval = rangeIntervals[0]
 	if rangeInterval > i.interval {
-		return rangeInterval, true
+		return rangeInterval, true, nil
 	}
 
 	level.Debug(i.logger).Log("msg", "unable to split expression because range interval is smaller than configured split interval", "expr", expr, "range_interval", rangeInterval, "split_interval", i.interval)
-	return time.Duration(0), false
+	return time.Duration(0), false, nil
 }
 
-// getRangeInterval returns the range interval in the range vector expression
-// Returns 0 if no range interval is found
-// Example: expression `count_over_time({app="foo"}[10m])` returns 10m
-func getRangeInterval(expr parser.Expr) time.Duration {
-	switch e := expr.(type) {
-	case *parser.AggregateExpr:
-		return getRangeInterval(e.Expr)
-	case *parser.Call:
-		// Iterate over Call's arguments until a MatrixSelector and a valid range interval are found
-		for _, arg := range e.Args {
-			if argRangeInterval := getRangeInterval(arg); argRangeInterval != 0 {
-				return argRangeInterval
-			}
+// getRangeIntervals recursively visit the input expr and returns a slice containing all range intervals found.
+func getRangeIntervals(expr parser.Expr) []time.Duration {
+	// Due to how this function is used, we expect to always find at most 1 range interval
+	// so we preallocate it accordingly.
+	ranges := make([]time.Duration, 0, 1)
+
+	// Ignore the error since we never return it.
+	_, _ = anyNode(expr, func(entry parser.Node) (bool, error) {
+		matrix, ok := entry.(*parser.MatrixSelector)
```

Comment:
If I understand correctly, this will return an empty array if `expr` is a subquery right?
If so, can we add a test to `TestGetRangeIntervals`?

Reply:
From looking at Suquery it can have an expression: https://github.com/prometheus/prometheus/blob/main/promql/parser/ast.go#L131-L146
so that means that we can have something like `sum_over_time(rate(metric[1m])[5h:5m])`, right? 
In this case, I think this will return `1m` and it cannot be split 🤔

Reply:
That's a good point! I think the utilities should be agnostic, but we should avoid splitting subqueries in the caller. I'm proposing it here: https://github.com/grafana/mimir/pull/2575/commits/37076ddd899cf00d8aedc6bbac02d36e93fc1403

We do something similar (but a bit more evoluted to support some subqueries) in query sharding:
https://github.com/grafana/mimir/blob/cc0ef427c9c2e516d844479028c2c15a726be740/pkg/frontend/querymiddleware/astmapper/sharding.go#L95

Reply:
Fixed here:
https://github.com/grafana/mimir/pull/2575/commits/1e88552d09a80e89cb2e6d714641e4e6719ab8f7